### PR TITLE
Publish HNSW adjacency as uint32_list vector-index state

### PR DESCRIPTION
**Pull request by snissn**
## Objective

Publish HNSW adjacency as first-class vector-index state assets using generic typed-column `uint32_list` / `raw_uint32_offsets_list` encoding. Review follow-ups tighten fail-closed asset presence checks, avoid full payload decode, reuse a scratch buffer across layers, and keep `VectorIndexStatus` on the cheap manifest-level validation path while deferring full payload validation to open.

## Context

This change is needed to make adjacency durable as typed vector-index state ahead of follow-up consumption work in #1988. The follow-up fixes ensure the typed adjacency contract is enforced consistently, keep loaded-status checks allocation-conscious for large layered graphs, and still fail closed when payload corruption is encountered during open.

## Non-goals

- Switching search consumption to read `TVIS` adjacency assets.
- Publishing normalized vectors or inverse-norm state.
- Removing legacy `column_graph` adjacency-source compatibility in this PR.

## Design

- Formats/flags/APIs changed (include diagrams if applicable)
  - Adds one `TVIS` `adjacency` asset per HNSW layer with asset IDs `hnsw/layer/<n>`.
  - Uses logical `uint32_list` and physical `raw_uint32_offsets_list`.
  - Keeps external search APIs unchanged.
  - Splits validation modes so status verifies asset availability, contiguous layer coverage, and schema/base identity cheaply, while open still performs full typed payload validation.
- Invariants that must hold
  - One adjacency asset per contiguous layer `0..maxLayer`.
  - `rows == graph row count`.
  - Offsets length is `rows+1`.
  - Final offset equals flattened neighbor count.
  - Neighbor ordinals are `< row_count`.
  - Base identity, schema hash, and asset refs must match.
- Fail-closed behavior (caps before alloc; CRC/error handling)
  - Status fails closed if adjacency assets are missing entirely, duplicated, gapped, unavailable, or schema/base identity mismatched.
  - Open fails closed on malformed typed payload structure, bad offsets/value counts, or out-of-bounds ordinals.
  - Full payload validation checks split-section metadata and byte shape before scanning values.
  - Ordinal validation scans raw offsets/value bytes directly instead of decoding full `[]uint64`/`[]uint32` slices.
  - Full adjacency validation reuses a single raw scratch buffer (`read...Into`) so open does not allocate a fresh full payload per layer.

## Scope

- Includes (files/symbols):
  - `TreeDB/collections/column_vector_index_state_adjacency.go`
  - `TreeDB/collections/column_vector_index_state_adjacency_test.go`
  - `TreeDB/collections/column_vector_index_state_manifest_test.go`
  - `TreeDB/collections/column_vector_graph_manifest.go`
  - `TreeDB/collections/typed_storage_naming_test.go`
  - `validateColumnVectorIndexStateAssets`
  - `validateColumnVectorIndexStateAssetsForStatus`
  - `validateColumnVectorIndexStateAdjacencyAssetInto`
- Excludes (files/symbols):
  - Searcher consumption switch for `TVIS` adjacency assets
  - Normalized vector or inverse-norm state publication
  - Removal of legacy `column_graph` adjacency-source readers/writers

## Correctness

- Tests run (exact commands):
  - `go test ./TreeDB/collections -run 'TestColumnGraphRebuildPublishesUint32ListAdjacencyState1987|TestColumnVectorIndexStateAdjacencyStatusValidation1987|TestColumnVectorIndexStatePublishedReopenAndFailClosed1986' -count=1`
  - `go test ./TreeDB/collections -count=1`
  - `GOWORK=off go test ./TreeDB/collections -count=1`
- Corruption/edge cases covered:
  - Missing adjacency assets
  - Missing layer 0
  - Missing intermediate adjacency layers
  - Duplicate adjacency layers
  - Out-of-bounds ordinals
  - Row-count mismatch
  - Final offset / value-count mismatch
  - Embedded schema-version mismatch
  - Stale base identity
  - Corrupt or unavailable asset refs
  - Cheap status loaded path with payload corruption deferred to open
- Concurrency/race coverage (if applicable):
  - Not applicable; this change is validation/publish logic only.

## Performance

- Benchmarks run (exact commands):
  - Existing PR benchmarks:
    - `go test ./TreeDB/collections -run '^$' -bench 'BenchmarkColumnGraph(RebuildVectorIndexV2A|VectorIndexStatusLoadedV2A)$' -benchmem -benchtime=3s -count=5`
    - `benchstat /tmp/gomap_1987_base_bench.txt /tmp/gomap_1987_head_bench_c046a72.txt`
  - No new benchmark was run for follow-up review-fix commits.
- Results table (before/after; median-of-5 per G1 in `docs/OPT_SPRINT_NEXT.md`):
  
  | Benchmark | main ns/op | PR ns/op | PR ops/sec | B/op | allocs/op | Notes |
  | --- | ---: | ---: | ---: | ---: | ---: | --- |
  | `ColumnGraphVectorIndexStatusLoadedV2A` | 59.23µ | 24.62µ | ~40,617 | ~7.63Ki | 106 | original PR head showed no alloc regression in loaded-status path |
  | `ColumnGraphRebuildVectorIndexV2A` | 17.39ms | 19.60ms | ~51.0 | 2.339Mi | 14.10k | rebuild alloc increase comes from publishing durable typed adjacency state |
- Regressions (if any) and why acceptable:
  - No new measured regression from follow-up fixes.
  - Full payload validation now avoids full decode and reuses a per-validation scratch buffer across layers.
  - Status no longer reads/validates full adjacency payload bytes, preserving the cheap loaded-status path and deferring heavy validation to open.

## Rollout / Toggle Plan

- Default setting:
  - Enabled by default for rebuilt column-graph vector indexes that publish typed adjacency state.
- How to disable quickly:
  - Rebuild behavior can be avoided by not consuming the typed state yet; there is no new runtime toggle in this PR.

## Follow-ups

- Issues/PRs to do next:
  - #1988 should consume `TVIS` adjacency assets by role `adjacency`, asset ID `hnsw/layer/<n>`, logical `uint32_list`, physical `raw_uint32_offsets_list`.
  - #1989 can remove or further isolate legacy graph adjacency compatibility once typed-state consumption lands.
  - #1992 owns normalized vectors / inverse-norm state.

### Copied/Adapted From Old Stack

- Copied:
  - None.
- Adapted:
  - Reused typed-column `raw_uint32_offsets_list` machinery for adjacency-state publication and validation.
- Oracle/comparator only:
  - Legacy `column_graph` adjacency-source compatibility remains as comparator/fallback context.
- Quarantined/not copied:
  - Graph-specific `adjacency_layout` and `ColumnStoreValueAdjacencyList` are not part of the new primary typed state contract.

### Path Identity

- Native reader:
  - Status validates typed adjacency state assets from `TVIS` via a cheap manifest-level path; open performs full typed payload validation before use.
- Decoded comparator:
  - Legacy graph manifest identity still gates whether state matches the graph.
- Legacy/native vector index:
  - Legacy graph assets are still published alongside the new typed state in this PR.
- Unsupported/fail-closed:
  - Missing adjacency assets, duplicate layers, layer gaps, bad row counts, bad offsets, bad ordinals, malformed typed payload metadata, and mismatched refs/schema/base identity fail closed.

### Base And Dependency State

- Base branch:
  - `main`
- Base commit:
  - `533f6062c80758fcbd9accb31fd9580257bd0ca2`
- Base PR/head:
  - PR head originally published as `c046a72ca314baf6a167c4e2066e0d21c103ad0c`; latest validated local head is `544f7842`.
- #1621/#1634 assumptions:
  - Existing manifest identity and fail-closed status/open behavior remain authoritative.
- Missing generic column-store primitives:
  - None for this PR; generic `uint32_list` / `raw_uint32_offsets_list` primitives are already in place.
- Parent review fixes propagated:
  - Yes; missing/gapped asset validation, raw-byte ordinal scanning (no full decode), scratch-buffer reuse across adjacency layers, and cheap status-vs-open validation splitting are included.

### Evidence

- Tests:
  - `GOWORK=off go test ./TreeDB/collections -count=1` passed on local head `544f7842`.
- Benchmarks:
  - Existing benchmark evidence from the original PR remains applicable; no fresh benchmark run for review-fix commits.
- Status/fallback proof:
  - Deterministic tests cover missing assets and layer gaps, plus corrupt/mismatch reopen/status cases; payload corruption now explicitly preserves loaded status and fails on open.
- Performance attribution:
  - Original rebuild alloc increase is from additional typed adjacency assets; follow-up fixes reduce validation-side allocation pressure and keep status off the full payload-read path.
- Local regressions found/fixed:
  - Updated state-status expectations so malformed adjacency payloads are rejected on open without forcing expensive payload validation during status.

### Test Plan Start

- Milestone tasks targeted:
  - Publish typed adjacency state and validate it robustly on status/open.
- Tests to add before or with implementation:
  - Publish/reopen validation and corruption cases for typed adjacency assets.
- Existing tests to preserve:
  - Existing column-graph rebuild and vector-index state reopen/status coverage.
- #1646 test-list changes made before implementation:
  - None.

### Performance Plan Start

- Relevant benchmarks/metrics for this PR:
  - `BenchmarkColumnGraphRebuildVectorIndexV2A`
  - `BenchmarkColumnGraphVectorIndexStatusLoadedV2A`
- Baseline/comparator command or reason not applicable:
  - Compared against `origin/main` using the benchmark commands above.
- Expected setup/search/materialization boundary:
  - Rebuild pays to materialize typed adjacency state; search consumption remains unchanged in this PR.
- Upstream costs explicitly out of scope:
  - Searcher migration to typed adjacency state.
- Local performance risks to watch:
  - Validation-side allocation or scanning overhead on large adjacency payloads, especially on status/open paths.

### Test Plan Close

- Additional tests found during implementation/review:
  - Missing adjacency assets
  - Missing layer 0
  - Missing intermediate layer gap
  - Payload corruption cases that should keep status loaded but fail on open
- Tests added or changed:
  - Added/updated `column_vector_index_state_adjacency_test.go`
  - Updated `column_vector_index_state_manifest_test.go` expectation setup
  - Updated `typed_storage_naming_test.go` allowlist coverage
- Failing tests fixed:
  - Aligned `TestColumnGraphVectorIndexStatusValidatesVectorIndexState1986` with the new typed-adjacency requirement.
  - Adjusted adjacency corruption expectations to match cheap status validation and fail-closed open behavior.
- #1646 test-list changes made at close:
  - None.
- Residual untested gaps, if any:
  - No dedicated post-review benchmark rerun for the status/open validation split.

### Performance Plan Close

- Benchmark commands run:
  - Original PR benchmark commands listed above; no new benchmark command was run for review-fix commits.
- Results summary with throughput/latency/allocations:
  - Original PR benchmark summary unchanged; review fixes target validation allocation behavior rather than rebuild/search throughput.
- Before/after or baseline comparison:
  - Existing comparison remains `origin/main` vs original PR head.
- Local slow/heavy code found and fixed:
  - Replaced eager full adjacency decode during validation with section validation + raw-byte ordinal scanning, reused validation scratch buffer across layers, and kept status on a manifest/schema validation path that avoids full payload reads.
- Remaining upstream or deferred costs:
  - Rebuild still pays the expected cost of publishing typed adjacency assets until follow-up cleanup/removal work lands.

### AI Review Loop

- Codex latest-head review:
  - No additional latest-head findings incorporated.
- Copilot latest-head review:
  - Latest head `544f784` reviewed; no additional issues found.
- CodeRabbit latest-head review:
  - No additional latest-head findings incorporated.
- Review comments/threads resolved or explicitly dismissed:
  - Addressed actionable review feedback covering missing/gapped fail-closed checks, raw-byte validation without full decode, scratch-buffer reuse, legacy `column_graph` error context, and cheap status-path validation.
- Re-requested after final meaningful push:
  - Yes.

---

### Scope
- [x] Single, clear objective for this PR (not a bundle)
- [x] Explicit include list (files/symbols touched):
- [x] Explicit exclude list (files/symbols NOT touched):
- [x] No unwanted feature reintroduction (e.g. async slab writer, hard zones, ValueIndex) unless explicitly stated in the PR objective

### Safety / Correctness
- [x] Clear written-vs-durable semantics for any `*Sync` path touched
- [x] No new mmap usage on mutable/truncating files
- [x] All new length fields are capped before allocation (OOM-safe)
- [x] CRC/checksum behavior is fail-closed (clean error, no panic)
- [x] Any concurrency change has a defined state machine and invariants

### Tests
- [x] Added deterministic regression tests for the targeted failure mode
- [x] Tests avoid sleeps where possible (use latches/hooks)
- [x] Added corruption/edge-case tests (truncation, bad headers, invalid lengths) when format/parsing changes
- [ ] Ran locally:
  - [ ] `go test ./... -count=1`
  - [x] Any targeted packages/benchmarks:

### Benchmarks / Measurements
- [ ] Added or updated a benchmark relevant to the change
- [x] Reported results in PR description (before/after, command, machine)
- [ ] Included “incompressible” (worst-case) results if compression is involved

### Docs
- [ ] Updated `docs/OPT_SPRINT_NEXT.md` milestone status (if applicable)
- [x] Documented any new config knobs + defaults
- [x] Called out any format change in PR description (pre-alpha OK)

### Rollout / Toggle Plan (if behavior-affecting)
- [x] Safe defaults (feature off or conservative threshold)
- [ ] Clear knobs to disable/revert behavior quickly
- [ ] Observability: counters/logs to verify effectiveness

## Summary by CodeRabbit

* **New Features**
  * Vector indices now publish per-layer adjacency state assets and include them in on-disk storage accounting.

* **Bug Fixes**
  * Stronger validation and stricter error handling for missing, mismatched, or corrupt vector-index state; status reporting reflects state-aware storage bytes.

* **Tests**
  * Expanded test coverage for adjacency-state publish/validation, corruption scenarios, reopen behavior, and storage-byte expectations.

### CODE REVIEW

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

**Comment by snissn** [comment]
@coderabbitai review

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

**Comment by snissn** [comment]
@coderabbitai review

File: TreeDB/collections/column_vector_index_state_adjacency.go [hunk 1]
```diff
@@ -0,0 +1,479 @@
...
+		}
+		if previous, ok := seenAdjacencyLayers[layer]; ok {
+			return fmt.Errorf("collections: vector-index state duplicate adjacency layer=%d asset ids %q and %q", layer, previous, asset.AssetID)
+		}
+		seenAdjacencyLayers[layer] = asset.AssetID
+		if err := validateColumnVectorIndexStateAdjacencyAsset(rootDir, collection, cfg, def, state, asset, layer); err != nil {
+			return fmt.Errorf("collections: vector-index state adjacency layer %d asset %q: %w", layer, asset.AssetID, err)
+		}
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_vector_index_state_adjacency.go]: Fixed in c9d141956: `validateColumnVectorIndexStateAssets` now fails closed when no adjacency assets are present, requires a contiguous layer set, and compares it to `graph.AdjacencyLayerCount` (with only a zero-count fallback). Missing/gapped-layer tests cover this. Resolving.

File: TreeDB/collections/column_vector_index_state_adjacency.go [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+	if err != nil {
+		return err
+	}
+	list, err := typedcolumn.DecodeRawUint32OffsetsListFallback(nil, nil, offsetsRaw, valuesRaw, state.RowCount)
+	if err != nil {
+		return err
+	}
+	if err := validateColumnVectorIndexStateAdjacencyList(layer, list); err != nil {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_vector_index_state_adjacency.go:468]: Fixed before the current head: status/open validation no longer materializes the full offsets-list with `DecodeRawUint32OffsetsListFallback`; it validates the raw sections and scans `offsetsRaw`/`valuesRaw` directly for ordinal bounds. Resolving.

File: TreeDB/collections/column_vector_index_state_adjacency.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_vector_index_state_adjacency.go:409]: Fixed in c9d141956: the validator requires adjacency `uint32_list` assets, checks the contiguous set, and fails closed when required layers are absent. Resolving.

File: TreeDB/collections/column_vector_index_state_adjacency.go [hunk 3]
```diff
@@ -0,0 +1,507 @@
...
+	if len(seenAdjacencyLayers) == 0 {
+		return errors.New("collections: vector-index state missing adjacency assets")
+	}
+	maxLayer := 0
+	for layer := range seenAdjacencyLayers {
+		if layer > maxLayer {
+			maxLayer = layer
+		}
+	}
+	for layer := 0; layer <= maxLayer; layer++ {
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_vector_index_state_adjacency.go]: Fixed in c9d141956: adjacency validation now receives the graph manifest snapshot and requires the published layers to match `graph.AdjacencyLayerCount`, so a truncated layer set is rejected. Resolving.

File: TreeDB/collections/vector_index_rebuild.go [hunk 4]
```diff
@@ -748,6 +748,18 @@ func prepareColumnVectorGraphRebuildManifest(collection string, cfg ColumnStoreC
...
 	}
+	statePartID := nextColumnVectorGraphPartIDAfter(prepared.Ref.PartID, prepared.Ref.PartID)
+	if len(prepared.AdjacencyLayerSources) > 0 {
+		for _, source := range prepared.AdjacencyLayerSources {
+			statePartID = nextColumnVectorGraphPartIDAfter(statePartID, source.Ref.PartID)
+		}
+	} else if prepared.Layer0AdjacencySource.Present {
+		statePartID = nextColumnVectorGraphPartIDAfter(statePartID, prepared.Layer0AdjacencySource.Ref.PartID)
+	}
+	stateAdjacencyAssets, err := prepareColumnVectorIndexStateAdjacencyAssets(assetRootDir, collection, cfg, def, manifest.Generation, statePartID, rows)
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/vector_index_rebuild.go:760]: Fixed in cd6148f4: loaded column-graph status now uses `columnVectorGraphStorageBytesWithState`, which adds `columnVectorIndexStateAssetsStorageBytes(state)` to the graph/source storage total. Tests now assert BytesDisk includes state assets, and the rebuild benchmark reports `state_assets_B/op` plus updated `graph_total_storage_B/op`.

File: TreeDB/collections/column_vector_index_state_adjacency.go [hunk 5]
```diff
@@ -0,0 +1,514 @@
...
+	if asset.SourceSchemaHash != sourceCfg.SchemaHash {
+		return fmt.Errorf("schema_hash=%d want %d", asset.SourceSchemaHash, sourceCfg.SchemaHash)
+	}
+	raw, err := readColumnPhysicalAssetFromManager(rootDir, asset.Ref)
+	if err != nil {
+		return err
+	}
+	if int64(len(raw)) != asset.AssetBytes || int64(len(raw)) != asset.Ref.Length {
+		return fmt.Errorf("bytes=%d manifest=%d ref=%d", len(raw), asset.AssetBytes, asset.Ref.Length)
+	}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_vector_index_state_adjacency.go]: Fixed in cd6148f4: `validateColumnVectorIndexStateAssets` now carries a reusable `rawScratch` buffer across adjacency layers and `validateColumnVectorIndexStateAdjacencyAssetInto` reads via `readColumnPhysicalAssetFromManagerInto`, so status/open validation reuses capacity up to the max single-layer payload instead of allocating a fresh full payload per layer.

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

**Comment by snissn** [comment]
@coderabbitai review

File: TreeDB/collections/column_vector_index_state_adjacency.go [hunk 6]
```diff
@@ -0,0 +1,522 @@
...
+		return rawScratch, err
+	}
+	if int64(len(raw)) != asset.AssetBytes || int64(len(raw)) != asset.Ref.Length {
+		return raw, fmt.Errorf("bytes=%d manifest=%d ref=%d", len(raw), asset.AssetBytes, asset.Ref.Length)
+	}
+	image, err := typedcolumn.ParseColumnPartImage(raw)
+	if err != nil {
+		return raw, err
+	}
+	if image.PartID != asset.Ref.PartID || image.Rows != asset.RowCount || image.Rows != state.RowCount {
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_vector_index_state_adjacency.go:440]: Fixed in fb211e2: state adjacency validation now validates the parsed typed-column image against the expected adapter schema via `typedColumnAdapterPartFromImageWithoutRowLocators` with `SchemaVersion: uint32(sourceCfg.SchemaHash)` before accepting offsets/list sections. Added `embedded_schema_version_mismatch` fail-closed coverage. Focused/package tests and the status/rebuild benchmark were rerun.

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

**Comment by snissn** [comment]
@coderabbitai review

File: TreeDB/collections/column_vector_index_state_adjacency_test.go [hunk 7]
```diff
@@ -0,0 +1,471 @@
...
+				return
+			}
+			t.Fatalf("descriptor section missing")
+		})
+		ctx.state.Assets[0] = bad
+		ctx.records, ctx.identity = appendVectorIndexStateRecordForTest1986(t, *ctx.cfg, ctx.records, ctx.identity, ctx.state)
+		publishColumnVectorIndexStateAdjacencyContext1987(t, d, ctx)
+		assertColumnVectorIndexStateAdjacencyStatusCorrupt1987(t, d, ctx.def, "schema_version")
+	})
+
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_vector_index_state_adjacency_test.go:186]: Fixed in 934d2c3: added a comment documenting that descriptor offset 10 targets the typed-column `schema_version` uint32 before mutating it.

File: TreeDB/collections/column_vector_graph_adjacency_source.go [hunk 8]
```diff
@@ -168,53 +168,9 @@ func prepareColumnVectorGraphAdjacencySourcePayloads(assetRootDir, collection st
...
-			}
-		}
-		for layer := range layers {
-			layers[layer].Offsets[rowIdx+1] = uint64(len(layers[layer].Values))
-		}
+	layers, err := buildColumnVectorIndexStateAdjacencyLists(rows)
+	if err != nil {
+		return nil, err
 	}
 	return layers, nil
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_vector_graph_adjacency_source.go:175]: Fixed in 934d2c3: `buildColumnVectorGraphAdjacencyLayerLists` now wraps delegated list-build failures with `collections: column_graph adjacency source layer lists`, so legacy adjacency-source publication errors keep column_graph context.

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

**Comment by snissn** [comment]
@coderabbitai review

File: TreeDB/collections/column_vector_graph_manifest.go [hunk 9]
```diff
@@ -888,12 +889,13 @@ func (c *Collection) columnGraphVectorIndexStatusAtSnapshot(name string, snap *b
 				status.RebuildNeeded = true
 				return status, nil
 			}
-			if err := validateColumnVectorIndexStateAssetRefsAvailable(c.db.ColumnAssetRootDir(), state); err != nil {
+			if err := validateColumnVectorIndexStateAssets(c.db.ColumnAssetRootDir(), catalog.meta.Name, *cfg, def, state, graph); err != nil {
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_vector_graph_manifest.go]: Fixed in c4b4b2c and carried through 76f6173 on top of main: `VectorIndexStatus` now returns rebuild-needed/asset-mismatch instead of loaded when the vector-index state record is absent, and `openColumnVectorGraphPhysicalRowReader` fails closed on a missing state record. Added `TestColumnGraphVectorIndexStatusRequiresVectorIndexStateRecord1987`; package tests pass after rebasing/merging main with #1992.

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

**Comment by snissn** [comment]
@coderabbitai review

File: TreeDB/collections/column_vector_graph_manifest.go [hunk 10]
```diff
@@ -888,7 +889,7 @@ func (c *Collection) columnGraphVectorIndexStatusAtSnapshot(name string, snap *b
...
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_vector_graph_manifest.go]: Fixed in 544f784: routine `VectorIndexStatus` now calls `validateColumnVectorIndexStateAssetsForStatus`, which only checks manifest identity, required contiguous adjacency layer coverage, schema hashes, and asset-ref availability. Full typed-column payload parsing and neighbor scans remain in `validateColumnVectorIndexStateAssets` for row-reader/open validation. The payload-corruption tests now assert cheap status can report loaded while open fails closed. Tests and status/rebuild benchmark rerun; artifacts under `/tmp/gomap_1987_tests_packages_after_status_cheap2.txt` and `/tmp/gomap_1987_benchstat_after_status_cheap.txt`.

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

**Comment by snissn** [comment]
@coderabbitai review

File: TreeDB/collections/column_vector_index_state_adjacency.go [hunk 11]
```diff
@@ -0,0 +1,543 @@
...
+	return assets
+}
+
+func validateColumnVectorIndexStateAssets(rootDir, collection string, cfg ColumnStoreConfig, def VectorIndexDefinition, state columnVectorIndexStateSnapshot, graph columnVectorGraphManifestSnapshot) error {
+	return validateColumnVectorIndexStateAssetsWithMode(rootDir, collection, cfg, def, state, graph, true)
+}
+
+func validateColumnVectorIndexStateAssetsForStatus(rootDir, collection string, cfg ColumnStoreConfig, def VectorIndexDefinition, state columnVectorIndexStateSnapshot, graph columnVectorGraphManifestSnapshot) error {
+	return validateColumnVectorIndexStateAssetsWithMode(rootDir, collection, cfg, def, state, graph, false)
+}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_vector_index_state_adjacency.go:355]: Fixed in 1d4b4d4: status validation is split from open validation. `validateColumnVectorIndexStateAssetsForStatus` now still reads/checks typed-column payload schema/version, row count, sections, offsets monotonic/final-offset shape and checksum/length, but skips only the O(total edges) neighbor ordinal scan. Open/reader validation keeps the full ordinal scan. Focused/package tests and status/rebuild benchmark rerun; artifacts: `/tmp/gomap_1987_tests_focused_after_status_split.txt`, `/tmp/gomap_1987_tests_packages_after_status_split.txt`, `/tmp/gomap_1987_benchstat_after_status_split.txt`.

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

**Comment by snissn** [comment]
@coderabbitai review

File: TreeDB/collections/column_vector_index_state_adjacency.go [hunk 12]
```diff
@@ -0,0 +1,544 @@
...
+		}
+		sourceCfg, _, err := columnVectorIndexStateAdjacencyColumnStoreConfig(collection, cfg, def, layer)
+		if err != nil {
+			return err
+		}
+		if asset.SourceSchemaHash != sourceCfg.SchemaHash {
+			return fmt.Errorf("collections: vector-index state adjacency layer %d schema_hash=%d want %d", layer, asset.SourceSchemaHash, sourceCfg.SchemaHash)
+		}
+		var validateErr error
+		rawScratch, validateErr = validateColumnVectorIndexStateAdjacencyAssetInto(rootDir, collection, cfg, def, state, asset, layer, validateOrdinals, rawScratch)
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_vector_index_state_adjacency.go]: Fixed in ed8d6fd: `validateColumnVectorIndexStateAssetsForStatus` is back to manifest/ref-only validation. It checks asset ref availability, role/asset id layer parsing, duplicate/contiguous layer coverage, expected layer count, and `SourceSchemaHash`, but no longer calls `validateColumnVectorIndexStateAdjacencyAssetInto` or reads/parses adjacency payload bytes. Heavy typed-column payload/schema/offset/ordinal validation remains in `validateColumnVectorIndexStateAssets` for open/rebuild paths. Focused tests and status benchmark rerun: `/tmp/gomap_1987_tests_focused_after_status_manifest_only.txt`, `/tmp/gomap_1987_tests_packages_after_status_manifest_only.txt`, `/tmp/gomap_1987_benchstat_after_status_manifest_only.txt`.

File: TreeDB/collections/column_vector_index_state_adjacency.go [hunk 13]
```diff
@@ -0,0 +1,544 @@
...
+		if validateErr != nil {
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_vector_index_state_adjacency.go]: Fixed in ed8d6fd: `validateColumnVectorIndexStateAssetsForStatus` is back to manifest/ref-only validation. It checks asset ref availability, role/asset id layer parsing, duplicate/contiguous layer coverage, expected layer count, and `SourceSchemaHash`, but no longer calls `validateColumnVectorIndexStateAdjacencyAssetInto` or reads/parses adjacency payload bytes. Heavy typed-column payload/schema/offset/ordinal validation remains in `validateColumnVectorIndexStateAssets` for open/rebuild paths. Focused tests and status benchmark rerun: `/tmp/gomap_1987_tests_focused_after_status_manifest_only.txt`, `/tmp/gomap_1987_tests_packages_after_status_manifest_only.txt`, `/tmp/gomap_1987_benchstat_after_status_manifest_only.txt`.

File: TreeDB/collections/column_vector_graph_manifest.go [hunk 14]
```diff
@@ -935,10 +937,20 @@ func (c *Collection) columnGraphVectorIndexStatusAtSnapshot(name string, snap *b
...
+		return status, nil
+	}
 	// Certified adjacency sources are optional accelerators. Keep loaded-status
 	// gating tied to the canonical row-asset graph; search opens and validates
 	// sources independently and falls back to row assets when they are absent,
 	// corrupt, stale, incomplete, or non-certified.
+	bytesDisk := columnVectorGraphStorageBytes(graph)
+	if loadedState != nil {
+		bytesDisk = columnVectorGraphStorageBytesWithState(graph, *loadedState)
+	}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_vector_graph_manifest.go]: Fixed in ed8d6fd: after the earlier `loadedState == nil` guard, `BytesDisk` now assigns `columnVectorGraphStorageBytesWithState(graph, *loadedState)` directly instead of retaining the redundant nil branch.

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

**Comment by snissn** [comment]
@coderabbitai review
